csrc/fusion_profiler.cpp

```diff
@@ -0,0 +1,770 @@
+
+  // Peak bandwidth calculation:
+  // Bus width is given in bits, so dividing by 8 converts to bytes.
+  // Clock is given in kHz. 1 GB = 1e9 bytes (don't report GiB = 1024^3 bytes)
+  // A factor of 2 is multiplied to account for double data rate (DDR):
+  // (clock in kHz * width in bits) * (1000 Hz / kHz) * (1 GB / 8e9 bits) * 2
+  // factor = 2.5e-7
+  constexpr double static_comp = 2.0 * /*2x for DDR*/
+      1000.0 * /*kHz->Hz*/
+      (1.0 / 8.0) * /*bits->bytes*/
+      (1.0 / 1.0e9); /*Bytes->Gigabytes*/
+  desc.peak_bandwidth_gbs = static_comp *
+      static_cast<double>(desc.memory_clock) *
+      static_cast<double>(desc.bus_width);
+}
+
+SegmentProfiler::SegmentProfiler(uint32_t id, bool cupti_disabled)
+    : cupti_disabled_(cupti_disabled),
+      device_(-1),
+      segment_id_(id),
+      compile_timer_(),
+      input_bytes_(0),
+      output_bytes_(0),
+      kernel_profile_state_(ProfilerState::Ready) {}
+
+void SegmentProfiler::startCompile(int device) {
+  device_ = device;
+  compile_timer_.start();
+}
+
+void SegmentProfiler::stopCompile() {
+  compile_timer_.stop();
+}
+
+void SegmentProfiler::startKernel(int device) {
+  device_ = device;
+  NVF_CHECK(
+      kernel_profile_state_ == ProfilerState::Ready,
+      "ProfilerState is not Ready!",
+      kernel_profile_state_);
+  if (!cupti_disabled_) {
+    NVFUSER_CUPTI_SAFE_CALL(cuptiActivityPushExternalCorrelationId(
+        CUPTI_EXTERNAL_CORRELATION_KIND_UNKNOWN,
+        static_cast<uint64_t>(segment_id_)));
+  }
+  kernel_profile_state_ = ProfilerState::Running;
+}
+
+void SegmentProfiler::stopKernel() {
+  NVF_CHECK(
+      kernel_profile_state_ == ProfilerState::Running,
+      "ProfilerState is not Running!",
+      kernel_profile_state_);
+  uint64_t corr_id = 0;
+  if (!cupti_disabled_) {
+    NVFUSER_CUPTI_SAFE_CALL(cuptiActivityPopExternalCorrelationId(
+        CUPTI_EXTERNAL_CORRELATION_KIND_UNKNOWN, &corr_id));
+    NVF_CHECK(
+        corr_id == static_cast<uint64_t>(segment_id_),
+        "Correlation Id does not match segment id! Corr Id: ",
+        corr_id,
+        " Segment Id: ",
+        segment_id_);
+  }
+  kernel_profile_state_ = ProfilerState::Finished;
+}
+
+void SegmentProfiler::inputBytesAccessed(int64_t bytes) {
+  input_bytes_ = bytes;
+}
+
+void SegmentProfiler::outputBytesAccessed(int64_t bytes) {
+  output_bytes_ = bytes;
+}
+
+uint32_t SegmentProfiler::segmentId() const {
+  return segment_id_;
+}
+
+void FusionProfile::reset() {
+  fusion_id = -1;
+  segments = 0;
+
+  cuda_evt_time_ms = 0.0;
+  host_time_ms = 0.0;
+  compile_time_ms = 0.0;
+  kernel_time_ms = 0.0;
+
+  input_bytes = 0;
+  output_bytes = 0;
+
+  effective_bandwidth_gbs = 0.0;
+  percentage_peak_bandwidth = 0.0;
+
+  kernel_profiles.clear();
+}
+
+std::array<const char*, 25> column_strs{
+    "Fus#",
+    "NSegs",
+    "CuEvtTm(ms)",
+    "HstTm(ms)",
+    "CmpTm(ms)",
+    "KerTm(ms)",
+    "EffBw(GB/s)",
+    "%PeakBw",
+    "Seg#",
```

Comment:
Keeping with the theme,
```suggestion
    "S-Seg#",
```